core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.join;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.units.DataSize;
+import io.trino.operator.HashArraySizeSupplier;
+import io.trino.operator.PagesHashStrategy;
+import io.trino.spi.Page;
+import io.trino.spi.PageBuilder;
+import io.trino.spi.block.Block;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.airlift.units.DataSize.Unit.KILOBYTE;
+import static io.trino.operator.SyntheticAddress.decodePosition;
+import static io.trino.operator.SyntheticAddress.decodeSliceIndex;
+import static io.trino.util.HashCollisionsEstimator.estimateNumberOfHashCollisions;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+// This implementation assumes:
+// -There is only one join channel and it is of type bigint
+// -arrays used in the hash are always a power of 2.
```

Comment:
Suppose we want to preserve as much memory as possible. In that case, we could use an approach similar to `io.trino.operator.HashGenerator#getPartition` to find the hash table slot and not rely on the power of two hash table size.

Reply:
Outside of scope here. The `BigintPagesHash` is a close of the default one. Bigger changes may land in subsequent PRs.
BTW this PR is definitely not about saving memory. Saving memory on hash tables will usually result in performance regressions.

Reply:
I made some benchmarks and there are two conclusions:
- calculating hash bucket like in `HashGenerator#getPartition` is slightly slower than a simple bit mask. Slightly, but it is visible in benchmarks
- Having a too big hash table instead of the one perfectly size increases performance simply because the average load is usually smaller than the max load factor. So this is a simple tradeoff between performance and memory and there is no incentive to change it to match perfectly the load factor
I do like this way of thinking though